// @silyze/async-audio-format-ulaw/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_audio_stream_1 = require("@silyze/async-audio-stream");
function ulawToPcm16kHz(ulaw) {
    const decoded = require("alawmulaw").mulaw.decode(new Uint8Array(ulaw));
    const pcmBuffer = Buffer.alloc(decoded.length * 2);
    for (let i = 0; i < decoded.length; i++) {
        pcmBuffer.writeInt16LE(decoded[i], i * 2);
    }
    return pcmBuffer;
}
function pcm16ToUlaw(pcm) {
    const pcmArray = new Int16Array(pcm.length / 2);
    for (let i = 0; i < pcmArray.length; i++) {
        pcmArray[i] = pcm.readInt16LE(i * 2);
    }
    const encoded = require("alawmulaw").mulaw.encode(pcmArray);
    return Buffer.from(encoded);
}
class ULawFormat extends async_audio_stream_1.AudioFormat {
    #sampleRate;
    constructor(sampleRate) {
        super();
        this.#sampleRate = sampleRate;
    }
    get name() {
        return "ulaw";
    }
    get pcmSampleRate() {
        return this.#sampleRate;
    }
    encode(input) {
        return input.transform().map(pcm16ToUlaw);
    }
    decode(input) {
        return input.transform().map(ulawToPcm16kHz);
    }
}
exports.default = ULawFormat;
