// @silyze/async-audio-format-ulaw/lib/index.ts
import { AsyncReadStream } from "@mojsoski/async-stream";
import { AudioFormat } from "@silyze/async-audio-stream";

function ulawToPcm16kHz(ulaw: Buffer): Buffer {
  const decoded: Int16Array = require("alawmulaw").mulaw.decode(
    new Uint8Array(ulaw)
  );
  const pcmBuffer = Buffer.alloc(decoded.length * 2);

  for (let i = 0; i < decoded.length; i++) {
    pcmBuffer.writeInt16LE(decoded[i], i * 2);
  }

  return pcmBuffer;
}

function pcm16ToUlaw(pcm: Buffer): Buffer {
  const pcmArray = new Int16Array(pcm.length / 2);

  for (let i = 0; i < pcmArray.length; i++) {
    pcmArray[i] = pcm.readInt16LE(i * 2);
  }

  const encoded: Uint8Array = require("alawmulaw").mulaw.encode(pcmArray);
  return Buffer.from(encoded);
}

export default class ULawFormat extends AudioFormat {
  #sampleRate: number;

  constructor(sampleRate: number) {
    super();
    this.#sampleRate = sampleRate;
  }

  get name() {
    return "ulaw";
  }

  get pcmSampleRate(): number {
    return this.#sampleRate;
  }

  encode(input: AsyncReadStream<Buffer>) {
    return input.transform().map(pcm16ToUlaw);
  }

  decode(input: AsyncReadStream<Buffer>) {
    return input.transform().map(ulawToPcm16kHz);
  }
}
